nova/virt/vmwareapi/vmops.py

```diff
@@ -2029,6 +2059,67 @@ def finish_migration(self, context, migration, instance, disk_info,
+        storage_policy = self._get_storage_policy(instance.flavor)
+        allowed_ds_types = ds_util.get_allowed_datastore_types(
+            image_meta.properties.hw_disk_type)
+        datastore = ds_util.get_datastore(self._session, self._cluster,
+                                          self._datastore_regex,
+                                          storage_policy,
+                                          allowed_ds_types)
+        dc_info = self.get_datacenter_ref_and_name(datastore.ref)
+        folder = self._get_project_folder(dc_info, instance.project_id,
+                                          'Instances')
+
+        spec = vm_util.relocate_vm_spec(self._session.vim.client.factory,
+                                        res_pool=self._root_resource_pool,
+                                        folder=folder, datastore=datastore.ref)
+
+        # Iterate over the network adapters and update the backing
+        if network_info:
+            spec.deviceChange = []
+            vif_model = image_meta.properties.hw_vif_model
+            hardware_devices = self._session._call_method(
+                                                    vutil,
+                                                    "get_object_property",
+                                                    vm_ref,
+                                                    "config.hardware.device")
+            vif_infos = vmwarevif.get_vif_info(self._session,
+                                               self._cluster,
+                                               utils.is_neutron(),
+                                               vif_model,
+                                               network_info)
+            for vif_info in vif_infos:
+                device = vmwarevif.get_network_device(hardware_devices,
+                                                      vif_info['mac_address'])
+                if not device:
+                    LOG.warning("Network device with MAC %s was not found in "
+                                "the VM while building the Relocate spec.",
+                                vif_info["mac_address"], instance=instance)
+                    continue
+                # Update the network device backing
+                config_spec = self._session.vim.client.factory.create(
+                    'ns0:VirtualDeviceConfigSpec')
+                vm_util.set_net_device_backing(self._session.vim.client.factory,
+                                               device, vif_info)
+                config_spec.operation = "edit"
+                config_spec.device = device
+                spec.deviceChange.append(config_spec)
+
+        vm_util.relocate_vm(self._session, vm_ref, spec=spec)
+
+    def _detach_volumes(self, instance, block_device_info):
+        block_devices = driver.block_device_info_get_mapping(block_device_info)
+        for disk in block_devices:
+            self._volumeops.detach_volume(disk['connection_info'], instance)
```

Comment:
Does this detach the root-disk for boot-from-volume, too? Is this intended?

Reply:
Looking at `_relocate_vm()`, it must be intended. Otherwise, boot-from-volume instances would end up on ephemeral storage.